### PR TITLE
Rewrite checkbox(task list) extension

### PR DESCRIPTION
**Pull request by chenzhiguang**

### CODE REVIEW

File: lib/src/block_syntaxes/list_syntax.dart [hunk 1]
```diff
@@ -165,7 +155,7 @@ abstract class ListSyntax extends BlockSyntax {
         }
         // End the current list item and start a new one.
         endItem();
-        childLines.add('$checkBoxIndicator$restWhitespace$content');
+        childLines.add(parseTastListItem(restWhitespace + content));
```
Comment (timmaffett) [inline lib/src/block_syntaxes/list_syntax.dart:158]: would 
  `childLines.add(parseTastListItem('$restWhitespace$content'));`
be preferred by dart style guidelines [prefer_interpolation_to_compose_strings](https://dart-lang.github.io/linter/lints/prefer_interpolation_to_compose_strings.html)

**Comment by timmaffett** [comment]
other than that minor nit it LGTM

File: lib/src/block_syntaxes/list_syntax.dart (hunk 1 above)
Comment (chenzhiguang) [reply lib/src/block_syntaxes/list_syntax.dart]: Thanks! Fixed.

File: lib/src/block_syntaxes/unordered_list_syntax.dart [hunk 2]
```diff
@@ -10,6 +11,21 @@ class UnorderedListSyntax extends ListSyntax {
   @override
   RegExp get pattern => ulPattern;
 
+  @override
+  bool canParse(BlockParser parser) {
+    // Check if it matches `hrPattern`, otherwise it will produce an infinite
```
Comment (srawlins) [inline lib/src/block_syntaxes/unordered_list_syntax.dart:16]: Thanks for catching this; I experienced this but didn't figure out what was causing it.

**Review by srawlins — CHANGES_REQUESTED** [review]
Thanks this looks great! I think I stepped on your toes and just landed some conflicting tests and patterns. Rebase and we should be good.

File: lib/src/block_syntaxes/list_syntax.dart [hunk 3]
```diff
@@ -6,21 +6,20 @@ import '../ast.dart';
...
 class ListItem {
-  ListItem(this.lines);
+  const ListItem(
+    this.lines, {
+    this.taskListItemStatus,
+  });
 
-  bool forceBlock = false;
   final List<String> lines;
+  final TaskListItemStatus? taskListItemStatus;
```
[the remark this answers is not in the document]
Comment (chenzhiguang) [reply lib/src/block_syntaxes/list_syntax.dart]: Fixed

File: lib/src/block_syntaxes/list_syntax.dart [hunk 4]
```diff
@@ -55,22 +54,40 @@ abstract class ListSyntax extends BlockSyntax {
...
     void endItem() {
       if (childLines.isNotEmpty) {
-        items.add(ListItem(childLines));
+        items.add(ListItem(childLines, taskListItemStatus: taskListItemStatus));
         childLines = <String>[];
       }
     }
 
-    late Match? possibleMatch;
+    String parseTastListItem(String text) {
```
[the remark this answers is not in the document]
Comment (chenzhiguang) [reply lib/src/block_syntaxes/list_syntax.dart]: Fixed

File: lib/src/block_syntaxes/list_syntax.dart [hunk 5]
```diff
@@ -6,21 +6,20 @@ import '../ast.dart';
...
 }
 
-/// Invisible string used to placehold for an *unchecked* CheckBox.
-/// The character is Unicode Zero Width Space (U+200B).
-const indicatorForUncheckedCheckBox = '\u{200B}';
-
-/// Invisible string used to placehold for a *checked* CheckBox.
-/// This is 2 Unicode Zero Width Space (U+200B) characters.
-const indicatorForCheckedCheckBox = '\u{200B}\u{200B}';
+enum TaskListItemStatus { checked, unchecked }
```
[the remark this answers is not in the document]
Comment (chenzhiguang) [reply lib/src/block_syntaxes/list_syntax.dart]: Fixed

File: lib/src/block_syntaxes/list_syntax.dart [hunk 6]
```diff
@@ -55,22 +54,40 @@ abstract class ListSyntax extends BlockSyntax {
...
+        items.add(ListItem(childLines, taskListItemStatus: taskListItemStatus));
         childLines = <String>[];
       }
     }
 
-    late Match? possibleMatch;
+    String parseTastListItem(String text) {
+      final pattern = RegExp(r'^ {0,3}\[([ xX])\][ \t]');
+
+      if (!taskListParserEnabled || !pattern.hasMatch(text)) {
```
[the remark this answers is not in the document]
Comment (chenzhiguang) [reply lib/src/block_syntaxes/list_syntax.dart]: Fixed. I have learnt it, thanks!

File: lib/src/block_syntaxes/list_syntax.dart [hunk 7]
```diff
@@ -55,22 +54,40 @@ abstract class ListSyntax extends BlockSyntax {
...
+      return text.replaceFirstMapped(pattern, ((match) {
+        taskListItemStatus = match[1] == ' '
+            ? TaskListItemStatus.unchecked
+            : TaskListItemStatus.checked;
+
+        return '';
+      }));
+    }
+
+    late Match? match;
```
[the remark this answers is not in the document]
Comment (chenzhiguang) [reply lib/src/block_syntaxes/list_syntax.dart]: Fixed. I did not change the names in purpose, I added this checkbox feature based on the old list syntax. But yes. the `possible Match ` and `successfulMatch` are more reasonable.

**Comment by chenzhiguang** [comment]
rebased

File: test/extensions/unordered_list_with_checkboxes.unit [hunk 8]
```diff
@@ -52,12 +44,12 @@
...
 <ul class="contains-task-list">
-<li class="task-list-item"><input type="checkbox"></input>​zero</li>
-<li class="task-list-item"><input type="checkbox"></input>​one</li>
-<li class="task-list-item"><input type="checkbox"></input>​two</li>
-<li class="task-list-item"><input type="checkbox"></input>​three</li>
-<li class="task-list-item"><input type="checkbox"></input>​four</li>
+<li class="task-list-item"><input type="checkbox"></input>one</li>
+<li class="task-list-item"><input type="checkbox"></input>two</li>
+<li class="task-list-item"><input type="checkbox"></input>three</li>
+<li class="task-list-item"><input type="checkbox"></input>four</li>
```
Comment (chenzhiguang) [inline test/extensions/unordered_list_with_checkboxes.unit:52]: @srawlins I do not know how could it happen. in your files there are two invisible whitespaces after each `</input>`

File: lib/src/block_syntaxes/list_syntax.dart (hunk 7 above)
[the remark this answers is not in the document]
Comment (chenzhiguang) [reply lib/src/block_syntaxes/list_syntax.dart]: `successfulMatch` has been changed to `match` as you did

File: lib/src/block_syntaxes/unordered_list_with_checkbox_syntax.dart [hunk 9]
```diff
@@ -2,16 +2,9 @@
...
-class UnorderedListWithCheckBoxSyntax extends ListSyntax {
-  @override
-  RegExp get pattern => ulWithCheckBoxPattern;
-
-  @override
-  String get listTag => 'ul_with_checkbox';
-
-  const UnorderedListWithCheckBoxSyntax();
+/// Parses unordered lists with checkboxes.
+class UnOrderedListWithCheckboxSyntax extends UnorderedListSyntax {
```
Comment (srawlins) [inline lib/src/block_syntaxes/unordered_list_with_checkbox_syntax.dart:8]: One nit here: all just particular capitals 🤣 

Thanks for changing `CheckBox` to `Checkbox`, but also we should keep `Unordered` as is. It is one word.

File: test/extensions/unordered_list_with_checkboxes.unit [hunk 10]
```diff
@@ -52,12 +52,12 @@
 *    [ ] four
 *     [ ] five
 <<<
+<p>*[ ] zero</p>
```
[the remark this answers is not in the document]
Comment (srawlins) [reply test/extensions/unordered_list_with_checkboxes.unit:55]: Agreed! Thanks!

File: test/extensions/unordered_list_with_checkboxes.unit (hunk 8 above)
Comment (srawlins) [reply test/extensions/unordered_list_with_checkboxes.unit:60]: Looks good

**Review by srawlins — CHANGES_REQUESTED** [review]
One more tiny change

**Review by srawlins — APPROVED** [review]
🎉

**Comment by devoncarew** [comment]
It looks like this PR removes some symbols from the published API (from casual inspection, `indicatorForCheckedCheckBox` and `indicatorForUncheckedCheckBox`, published as part of v6.0.0; I didn't look to see if there were more API changes). There are packages using that API - currently at least package:flutter_markdown. I don't see a changelog entry, pubspec change, or version bump associated w/ this PR.

This is a breaking change; is that intended? Is it meant to ship as part of the current major version (v6.0.0)? Or as the next major version?

**Comment by srawlins** [comment]
Sorry we can re-introduce deprecated accessors for those before publishing

**Comment by timmaffett** [comment]
I will additionally add that those symbols were only used within list_syntax.dart (essentially privately and perhaps erroneously not prefixed with '_').  
It is probably unlikely that they were of use to anyone (or being used), but of course adding them back will not hurt anything and is the safe thing to do.  

(There were used as invisible markers within the text string (within list_syntax.dart) to indicate where a checkbox should be inserted.  Their use will no longer have any effect).

**Comment by devoncarew** [comment]
> I will additionally add that those symbols were only used within list_syntax.dart (essentially privately and perhaps erroneously not prefixed with '_'). It is probably unlikely that they were of use to anyone (or being used), but of course adding them back will not hurt anything and is the safe thing to do.

Yeah, it seems likely that they were not widely used; I did however find usages from flutter_markdown.

I could see one option being to add them back, but w/ a deprecation annotation, and to tell people that they'll be removed in the next major version.

A 2nd options would be to keep them removed, update some docs to indicate that the symbols were added to the namespace inadvertently, and update flutter_markdown to not use the symbols. That's not following semver - removing API within a major version - but in practice may not be very breaking.

**Comment by devoncarew** [comment]
For posterity, here is the current ecosystem usage of markdown's API (this is just from packages on pub.dev, not app usage):

https://gist.github.com/devoncarew/cfc3ac9f2efd3fe059e4149ea84a25b7

It does look like package:flutter_markdown is the only thing currently using the `indicatorFor*` symbols.

**Comment by timmaffett** [comment]
I looked at the usage within flutter_markdown, and it is within `list_test.dart` test.  They are used to create strings of the expected output for the test.   This test will need to be updated for the new library anyway - so this usage will be removed - the but following change would allow it to continue to pass unchanged.

If these are being used anywhere I suspect it is exactly like `list_test.dart` is using them.

I think the important thing is for the deprecated accessors we add back to return EMPTY STRINGS.  This will ensure that any usage similar to what is being done in list_test.dart will result in tests that still pass *without* any changes.  The reason for this is that the previous invisible non breaking spaces are no longer used, so now there would be nothing found in the output from markdown. Hence returning empty strings will allow the test to pass.

Essentially adding back in deprecated accessors something like this:
```
@Deprecated('This string is no longer used internally.  It will be removed in a future version.')
const indicatorForUncheckedCheckBox = '';
@Deprecated('This string is no longer used internally.  It be will be removed in a future version.')
const indicatorForCheckedCheckBox = '';
```


I will try this and test out the flutter_markdown test and see if they pass.  I will then completely remove their usage from the test and verify the changed test works correctly with the checkbox rewrite.
Trying the empty accessors first will allow us to verify that doing this will indeed allow similar test usage to pass without code changes.  (Essentially for our peace of mind that this is the correct approach).

**Comment by timmaffett** [comment]
I completed the changes and testing to verify they work as expected. 
1) re-addition of  (deprecated) `indicatorForUncheckedCheckBox` and `indicatorForCheckedCheckBox` symbols to markdown - changed to return empty strings.  (https://github.com/dart-lang/markdown/pull/459)
2) updating of the `list_test.dart` withing flutter_markdown to completely remove the use of these deprecated symbols. (https://github.com/flutter/packages/pull/2613)

Either 1) or 2) or *both* allow flutter_markdown to compile and for all test to pass.  I would expect that would be the same for any other users of these symbols that may exist, as it would likely be a similar use to `list_test.dart`'s.

I have submitted PR's of these changes to markdown (for 1) and flutter_markdown (for 2)

**Pull request merged**